File: python/construct_k_palindrome_strings.py
class Solution:
    def canConstruct(self, s: str, k: int) -> bool:
        N = len(s)
        if k == N:
            return True
        elif k > N:
            return False
        # The trick is: if any letter appears an odd number of times,
        # then you can only include it in a palindrome sandwiched
        # between other letters, e.g. "dad".
        # So, by the pigeonhole principle, if there's <= k letters
        # that appear an odd number of times, we're good.
        hist = collections.Counter(s)
        letters_with_odd_appearances = 0
        for v in hist.values():
            if v % 2 != 0:
                letters_with_odd_appearances += 1
            if letters_with_odd_appearances > k:
                return False
        return letters_with_odd_appearances <= k

File: c/construct_k_palindrome_strings.c
bool canConstruct(char* s, int k) {
    int i = 0;
    int N = strlen(s);

    if (N == k) {
        return true;
    } else if (k > N) {
        return false;
    }

    int counter[26];
    char current_letter;
    for (i = 0; i < N; i++) {
        current_letter = s[i] - 'a';
        counter[current_letter] += 1;
    }
    int odd_character_count = 0;
    for (i = 0; i < 26; i++) {
        if ((counter[i] % 2) != 0) {
            odd_character_count += 1;
        }
        if (odd_character_count > k) {
            return false;
        }
    }
    if (odd_character_count > k) {
        return false;
    } else {
        return true;
    }
}
